File: fathersDay/Dsa/Day 30 (98. Validate Binary Search Tree)/Solution.cpp
#include <iostream>
#include <limits>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isBST(root, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max());
    }

private:
    bool isBST(TreeNode* node, long long low, long long high) {
        if (!node) {
            return true;
        }
        if (node->val <= low || node->val >= high) {
            return false;
        }
        return isBST(node->left, low, node->val) && isBST(node->right, node->val, high);
    }
};

int main() {
    TreeNode* root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);
    Solution solution;
    bool result = solution.isValidBST(root);
    std::cout << (result ? "True" : "False") << std::endl; // Output: True
    return 0;
}

File: fathersDay/Dsa/Day 30 (98. Validate Binary Search Tree)/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    public boolean isValidBST(TreeNode root) {
        return isBST(root, null, null);
    }

    private boolean isBST(TreeNode node, Integer low, Integer high) {
        if (node == null) {
            return true;
        }
        if ((low != null && node.val <= low) || (high != null && node.val >= high)) {
            return false;
        }
        return isBST(node.left, low, node.val) && isBST(node.right, node.val, high);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);
        Solution solution = new Solution();
        boolean result = solution.isValidBST(root);
        System.out.println(result); // Output: true
    }
}
